### PR TITLE
Big Minigames Update

### PR DESCRIPTION
**Pull request by SecretAgent12**

### CODE REVIEW

**Comment by SecretAgent12** [comment]
1 problem left
idk how to handle destroying vehicles by explosions

**Comment by GoldSloth** [comment]
You should be able to find out the shooter vehicle/player if the explosion is caused by an explosion. I think that's probably required before I merge this, I think.

File: src/main/java/com/flansmod/common/PlayerHandler.java [hunk 1]
```diff
@@ -37,6 +38,8 @@
     private static final Random rand = new Random();
     public static Map<String, PlayerData> serverSideData = new HashMap<String, PlayerData>();
     public static Map<String, PlayerData> clientSideData = new HashMap<String, PlayerData>();
+    public static Map<String, PlayerInfo> serverSideInfo = new HashMap<String, PlayerInfo>();
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/PlayerHandler.java]: I think PlayerInfo needs to be renamed to "PlayerStatistics" or "PlayerTeamsStatistics", and change serverSideInfo to serverSidePlayerStats to make it clear what the difference to serverSideData is

File: src/main/java/com/flansmod/common/PlayerHandler.java [hunk 2]
```diff
@@ -139,6 +142,26 @@ public static PlayerData getPlayerData(String username, Side side) {
         return side.isClient() ? clientSideData.get(username) : serverSideData.get(username);
     }
 
+    //---
+
+    public static PlayerInfo getPlayerInfo(EntityPlayerMP player) {
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/PlayerHandler.java]: likewise with this, getPlayerStatistics

File: src/main/java/com/flansmod/common/driveables/EntityDriveable.java [hunk 3]
```diff
@@ -449,6 +449,12 @@ public boolean isMountedEntity(Entity entity) {
 
     @Override
     public void setDead() {
+
+        TeamsManager teamsManager = TeamsManager.instance;
+        if(teamsManager.currentRound!=null){
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/driveables/EntityDriveable.java]: What's this code doing?

File: src/main/java/com/flansmod/common/network/PacketGetPlayerClasses.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    public void decodeInto(ChannelHandlerContext ctx, ByteBuf data) {
+
+    }
+
+    @Override
+    public void handleServerSide(EntityPlayerMP playerEntity) {
+        System.out.println("recievedGPC");
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/network/PacketGetPlayerClasses.java]: This needs to be removed

File: src/main/java/com/flansmod/common/teams/CommandTeams.java [hunk 5]
```diff
@@ -267,15 +267,34 @@ public void processCommand(ICommandSender sender, String[] split) {
             sender.addChatMessage(new ChatComponentText("Autobalance is now " + (TeamsManager.autoBalance ? "enabled" : "disabled")));
             return;
         }
-        if (split[0].equals("useRotation")) {
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/teams/CommandTeams.java]: It looks like you've removed the useRotation command, it needs to be added back

File: src/main/java/com/flansmod/common/teams/GameType.java [hunk 6]
```diff
@@ -128,6 +132,42 @@ public boolean playerCanAttack(EntityPlayerMP attacker, Team attackerTeam, Entit
...
+        if (bullet!=null && driveable2!=null) {
+            if (bullet.owner != null) {
+                EntityPlayerMP attacker = (EntityPlayerMP) bullet.owner;
+                EntityDriveable driveable = driveable2;
+//                if(driveable.riddenByEntity!=null &&
+//                        driveable.riddenByEntity instanceof EntityPlayer &&
+//                        !getPlayerData((EntityPlayerMP) driveable.riddenByEntity).team.equals(getPlayerData(attacker).team)) {
+                    if(true){
+                        getPlayerInfo(attacker).vehiclesDestroyed++;
+                        FlansMod.log("Minigames: "+driveable.getDriveableType().name + " destroyed by "+attacker+" with using "+bullet.firedFrom.name);
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/teams/GameType.java]: This should probably be removed too, I don't think it's a good idea to log every vehicle death.

File: src/main/java/com/flansmod/common/teams/GameTypeCTF.java [hunk 7]
```diff
@@ -3,6 +3,8 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import com.flansmod.common.PlayerHandler;
+import ibxm.Player;
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/teams/GameTypeCTF.java]: I don't think this import is used?

File: src/main/java/com/flansmod/common/teams/GameTypeTDM.java [hunk 8]
```diff
@@ -39,49 +80,49 @@ public void roundCleanup() {
 
     @Override
     public void tick() {
-        if (autoBalance && time % autoBalanceInterval == autoBalanceInterval - 200 && needAutobalance()) {
-            TeamsManager.messageAll("\u00a7fAutobalancing teams...");
-        }
-        if (autoBalance && time % autoBalanceInterval == 0 && needAutobalance()) {
-            autobalance();
-        }
+//        if (autoBalance && time % autoBalanceInterval == autoBalanceInterval - 200 && needAutobalance()) {
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/teams/GameTypeTDM.java:81]: er, why comment all of this out?

File: src/main/java/com/flansmod/common/teams/TeamsManager.java [hunk 9]
```diff
@@ -1012,6 +1012,7 @@ public void resetInventory(EntityPlayer player) {
         for (int i = 0; i < player.inventory.getSizeInventory(); i++) {
             ItemStack stack = player.inventory.getStackInSlot(i);
             if (stack != null && stack.getItem() instanceof ItemGun) {
+                new ItemStack(GunType.getGun("mtar").getItem());
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/teams/TeamsManager.java]: This needs to be removed

**Review by GoldSloth — DISMISSED** [review]
Good work so far, I've added quite a few questions and changes that need to be made, mainly just for clean code/readability. Let me know if you've got any questions.

File: src/main/java/com/flansmod/common/driveables/EntityDriveable.java (hunk 3 above)
Comment (SecretAgent12) [reply src/main/java/com/flansmod/common/driveables/EntityDriveable.java]: sorry, i forgot remove this

File: src/main/java/com/flansmod/common/PlayerHandler.java (hunk 1 above)
Comment (SecretAgent12) [reply src/main/java/com/flansmod/common/PlayerHandler.java]: okay

File: src/main/java/com/flansmod/common/PlayerHandler.java (hunk 2 above)
Comment (SecretAgent12) [reply src/main/java/com/flansmod/common/PlayerHandler.java]: okay

File: src/main/java/com/flansmod/common/teams/CommandTeams.java (hunk 5 above)
Comment (SecretAgent12) [reply src/main/java/com/flansmod/common/teams/CommandTeams.java]: sure

File: src/main/java/com/flansmod/common/teams/GameType.java (hunk 6 above)
Comment (SecretAgent12) [reply src/main/java/com/flansmod/common/teams/GameType.java]: okay

File: src/main/java/com/flansmod/common/teams/GameTypeTDM.java (hunk 8 above)
Comment (SecretAgent12) [reply src/main/java/com/flansmod/common/teams/GameTypeTDM.java:81]: I forgot to remove this code earlier.
  this is necessary for the normal operation of the autobalancer

File: src/main/java/com/flansmod/common/PlayerHandler.java (hunk 1 above)
Comment (SecretAgent12) [reply src/main/java/com/flansmod/common/PlayerHandler.java]: do I need to change all the files again?

File: src/main/java/com/flansmod/common/driveables/EntityDriveable.java [hunk 10]
```diff
@@ -1,2809 +1,233 @@
-package com.flansmod.common.driveables;
+package com.flansmod.common.teams;
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/driveables/EntityDriveable.java]: Something has gone seriously wrong here, why has EntityDriveable become the teams manager?

File: src/main/java/com/flansmod/common/PlayerHandler.java (hunk 1 above)
Comment (GoldSloth) [reply src/main/java/com/flansmod/common/PlayerHandler.java]: This has been resolved.

File: src/main/java/com/flansmod/common/driveables/EntityVehicle.java [hunk 11]
```diff
@@ -1,1349 +1,2802 @@
 package com.flansmod.common.driveables;
 
+import cofh.api.energy.IEnergyContainerItem;
```
Comment (GoldSloth) [inline src/main/java/com/flansmod/common/driveables/EntityVehicle.java]: What happenned to this file? Why is this one now EntityDriveable?

File: src/main/java/com/flansmod/common/teams/CommandTeams.java (hunk 5 above)
Comment (GoldSloth) [reply src/main/java/com/flansmod/common/teams/CommandTeams.java]: This has been resolved.

File: src/main/java/com/flansmod/common/driveables/EntityVehicle.java (hunk 11 above)
Comment (SecretAgent12) [reply src/main/java/com/flansmod/common/driveables/EntityVehicle.java]: when making changes, I mixed up the files

**Pull request merged**